### ISSUE TITLE
DNS timeout issues with v0.14.1

### ISSUE THREAD

**Issue by ciapecki** [dlenski/vpn-slice#54]
```
$ vpn-slice --version
vpn-slice 0.14.1
```

```
Connected as XXXX + YYYYYY/64, using SSL, with DTLS disabled
WARNING: IPv6 address or netmask set. Support for IPv6 in vpn-slice should be considered BETA-QUALITY.
WARNING: IPv6 address or netmask set. Support for IPv6 in vpn-slice should be considered BETA-QUALITY.
Blocked incoming traffic from VPN interface with iptables.
Added routes for 3 nameservers, 1 subnets, 0 aliases.
Restored routes for 0 excluded subnets. []
Adding /etc/hosts entries for 3 nameservers...
  XXX = dns0.tun0
  YYY = dns1.tun0
  ZZZZZZZZ = dns2.tun0
Looking up 55 hosts using VPN DNS servers...
WARNING: Lookup for xxx on VPN DNS servers failed:
	The DNS operation timed out after 30.00084924697876 seconds
WARNING: Lookup for yyy on VPN DNS servers failed:
	The DNS operation timed out after 30.000951528549194 seconds
WARNING: Lookup for zzz on VPN DNS servers failed:
	The DNS operation timed out after 30.000957012176514 seconds
WARNING: Lookup for sss on VPN DNS servers failed:
	The DNS operation timed out after 30.00096583366394 seconds
^CSend BYE packet: Aborted by caller
```

reverting back to v0.14 resolves these issues for me:
```
$ vpn-slice --version
vpn-slice 0.14
```
```
Connected as XXXX + YYYYYYYYY/64, using SSL, with DTLS disabled
WARNING: IPv6 address or netmask set, but this version of vpn-slice has only rudimentary support for them.
WARNING: IPv6 address or netmask set, but this version of vpn-slice has only rudimentary support for them.
Blocked incoming traffic from VPN interface with iptables.
Added routes for 3 nameservers, 1 subnets, 0 aliases.
Restored routes for 0 excluded subnets.
Adding /etc/hosts entries for 3 nameservers...
  XXX = dns0.tun0
  YYY = dns1.tun0
  ZZZZZZZZ = dns2.tun0
Looking up 55 hosts using VPN DNS servers...
[...]
Added hostnames and aliases for 98 addresses to /etc/hosts.
Added 92 routes for named hosts.
Connection setup done, child process 2644287 exiting.
```

There must have been changes between v0.14 and v0.14.1 causing above timeouts. 
Anything I can test on my side?

thanks,
Chris

**Comment by ciapecki** [comment]
I noticed this in --dump -vv
in master (v0.14.1):
```  INTERNAL_IP6_ADDRESS    => myaddr6=IPv6Interface('2606:b400:2b5b:88:8000::13b4/64')
  INTERNAL_IP6_NETMASK    => netmask6=IPv6Interface('2606:b400:2b5b:88:8000::13b4/64')
```
whereas in (v0.14):
```  INTERNAL_IP6_ADDRESS    => myaddr6=IPv6Address('2606:b400:2b5b:89:8000::196')
  INTERNAL_IP6_NETMASK    => netmask6=IPv6Interface('2606:b400:2b5b:89:8000::196/64')
```

**Comment by dlenski** [comment]
Hmmm… [here are all the changes](https://github.com/dlenski/vpn-slice/compare/v0.14..v0.14.1) between v0.14 and v0.14.1.

> ```
> Adding /etc/hosts entries for 3 nameservers...
>   XXX = dns0.tun0
>   YYY = dns1.tun0
>   ZZZZZZZZ = dns2.tun0
> ```

Am I to assume that XXX and YYY are IPv4 addresses, while ZZZZZZZZ is an IPv6 address?

Once connected, can you `ping`/`ping6` all of these addresses?

What happens if you try to do the DNS lookups "manually" via `dig @DNS_server_IP host.to.look.up`, after connecting?

**Comment by ciapecki** [comment]
> As for the netmask changes between v0.14 and v0.14.1… is this a _consistent and persistent_ difference? These values are sent by the server itself… so it's quite like that if the prefix changed from `2606:b400:2b5b:88/64` to `2606:b400:2b5b:89/64`, it's because the server "wanted it" to change.

I didn't mean the exact values, but the following:
```
v0.14:  INTERNAL_IP6_ADDRESS    => myaddr6=**IPv6Interface**(...
v0.14.1: INTERNAL_IP6_ADDRESS    => myaddr6=**IPv6Address**(...
```

I will try to check your other thoughts later.

**Comment by dlenski** [comment]
I've attempted to reproduce this, by adding bogus/black-holed DNS servers to my list…

```
from vpn_slice.dnspython import DNSPythonProvider
import ipaddress
p=DNSPythonProvider()

# One bogus IPv4 DNS server, two real IPv6 DNS servers
p.configure(list(map(ipaddress.ip_address, (
    '123.45.67.89', # bogus IPv4 DNS server
    '2001:4860:4860::8888','2001:4860:4860::8844', # real Google Public DNS servers (IPv6)
))), bind_addresses=list(map(ipaddress.ip_address, ('192.168.12.233','2601:…'))) )
p.lookup_host('www.google.com')
```

Can you test fdf9ef2bcae634cbadace905df569f17b528f7b0 and see if it fixes the problem for you?

> IPv6**Interface** vs IPv6**Address**

Ah, okay. This was an intended change, to correctly replicate the behavior of the "standard" vpnc-script in terms of how it handles the parameters sent by openconnect. If it had an incorrect effect on your (IPv6) routing configuration, I'm interested to hear it.

**Comment by gmacon** [comment]
I've started seeing this, too. I looked at the traffic with Wireshark, and I now suspect a bug in dnspython. I was seeing a request go out from dnspython, and an empty response come back from the name server immediately. If dnspython were correctly receiving the response, I'd expect `NoAnswer` to be raised right away.

There were two differences between a successful request from dig and the request from dnspython: dig sets the AD flag, and dig uses EDNS0. I reconfigured dnspython to use EDNS0, but that didn't fix the issue. I didn't see an obvious way to get dnspython to set the AD flag, but I wouldn't really expect that to change anything.

**Comment by dlenski** [comment]
> I've started seeing this, too. I looked at the traffic with Wireshark, and I now suspect a bug in dnspython. I was seeing a request go out from dnspython, and an empty response come back from the name server immediately.

Huh, that's interesting.

So your idea is that `dnspython` isn't raising `NoAnswer` where it should be… but is instead waiting and eventually timing out?

> There were two differences between a successful request from dig and the request from dnspython: dig sets the AD flag, …

Yeah, I don't see how that would change anything either.

> and dig uses EDNS0.

:thinking:
It's not _supposed_ to use EDNS… we disable that explicitly.

https://github.com/dlenski/vpn-slice/blob/75f2ee77996a288ac339b6e8ec1723dc327151d2/vpn_slice/posix.py#L14

**Comment by gmacon** [comment]
> So your idea is that dnspython isn't raising NoAnswer where it should be… but is instead waiting and eventually timing out?

That's right.

> It's not supposed to use EDNS… we disable that explicitly.

This one's my fault. I compared vpn-slice's traffic with running dig by hand, so just `dig @server_ip some.internal.name`. Sorry about that.

**Comment by dlenski** [comment]
>> So your idea is that dnspython isn't raising NoAnswer where it should be… but is instead waiting and eventually timing out?
>
> That's right.

I wish I could reproduce this, but cannot.

When I try fdf9ef2 (**with the logging statement uncommented**) on any of the real VPNs and DNS servers that I have access to (running on Ubuntu 18.04 here)… I get an immediate reply, rather than a timeout, end then the remaining DNS lookups complete successfully:

```
Issuing query for hostname 'fnord.nothing.net', rectype 'A', source IPv4Address('10.4.254.248'), search_domains [<DNS name foo.com.>, <DNS name foo.local.>], nameservers ['10.4.16.122', '10.4.16.166']
WARNING: Lookup for fnord.nothing.net on VPN DNS servers returned nothing.
Issuing query for hostname 'www.slashdot.org', rectype 'A', source IPv4Address('10.4.254.250'), search_domains [<DNS name foo.com.>, <DNS name foo.local.>], nameservers ['10.4.16.122', '10.4.16.166']
Got results: [<DNS IN A rdata: 216.105.38.15>]
  www.slashdot.org = 216.105.38.15
```

Could you try to use something like the [short standalone test script I posted above](https://github.com/dlenski/vpn-slice/issues/54#issuecomment-641482184) to repeatedly generate the DNS queries that are causing the timeout with `dnspython`?

**Comment by ciapecki** [comment]
> Can you test [fdf9ef2](https://github.com/dlenski/vpn-slice/commit/fdf9ef2bcae634cbadace905df569f17b528f7b0) and see if it fixes the problem for you?

I tested with your fix, and it works. It's not timing out, just resolving of the servers takes a long time before the connection is established.

**Comment by dlenski** [comment]
> I tested with your fix, and it works. It's not timing out, just resolving of the servers takes a long time before the connection is established.

Ahhhh… that's good to know.

Can you uncomment this additional logging statement and run again? https://github.com/dlenski/vpn-slice/commit/fdf9ef2bcae634cbadace905df569f17b528f7b0#diff-6213469e2b2c96b8c072c5abf4d783b7R35-R36

I'm _guessing_ that what's happening here is that it's timing out on the IPv6 DNS server(s) or the IPv4 DNS server(s) — whichever ones it tries first — and then after 30s tries on the _other_ IP-version servers, and succeeds.

If you uncomment the logging statement and try again, it will be 100% clear if this is what's happening.

**Comment by gmacon** [comment]
I'm running the version with extra logging, and it looks like I'm just impatient. It's successfully looking up the DNS on the IPv4 DNS server and then failing to do so against the IPv6 server (and there are, in fact, no IPv6 packets in Wireshark). Because I have a bunch of names, it'll take ten minutes to wait out all the IPv6 timeouts, and I never did.

I think it's surprising that, after successfully resolving a name against one name server, it tries again against the other. It strikes me that a configuration where AAAA records are only returned over IPv6 would be a very unusual one.

I'll try to debug my IPv6 connectivity.

**Comment by ciapecki** [comment]
> Can you uncomment this additional logging statement and run again? [fdf9ef2#diff-6213469e2b2c96b8c072c5abf4d783b7R35-R36](https://github.com/dlenski/vpn-slice/commit/fdf9ef2bcae634cbadace905df569f17b528f7b0#diff-6213469e2b2c96b8c072c5abf4d783b7R35-R36)

it works but still way too long. I tested for 5 servers it took 2,5 minutes to resolve.
on v0.14 it works almost immediately (couple of seconds)

**Comment by gmacon** [comment]
I just pushed a branch to my fork, [dns-by-route](https://github.com/gmacon/vpn-slice/tree/dns-by-route) that contains a modification so that dnspython can figure out what servers it wants to talk to on its own. This does remove the explicit source addresses, but they don't seem to be required for me. Thoughts?

**Comment by ciapecki** [comment]
@gmacon your fork works fine for me

**Comment by dlenski** [comment]
> it works but still way too long. I tested for 5 servers it took 2,5 minutes to resolve.
> on v0.14 it works almost immediately (couple of seconds)

Hmmm… 2.5 minutes is impossibly long with a 30 second timeout. There are at most 4 combinations of (DNS server address: IPv4 vs. IPv6 ) × (DNS rectype: `A` vs `AAAA`).

@ciapecki Can you share the logging output? It should not take more than 2 minutes given 4 combinations and a 30 second timeout.

**Comment by dlenski** [comment]
> … that contains a modification so that dnspython can figure out what servers it wants to talk to on its own. This does remove the explicit source addresses, but they don't seem to be required for me. Thoughts?

@gmacon thanks.

I can understand why this _works_ better, but I worry that it may go against important design goals of `vpn-slice`, namely:

1. not to leak anything about what you're doing "inside" the VPN to the rest of the Internet, and
2. not to leak anything about what you're doing on the rest of the Internet "inside" the VPN.

Not specifying the source-address means that `dnspython` may, for example, send some of your DNS queries out, unencrypted on the public Internet. This *shouldn't* happen _iff_ the routing tables _only_ have routes to the DNS servers via the VPN tunnel interface… but I'm not at all confident that all OSes handle it in this way.

**Comment by dlenski** [comment]
> I'm running the version with extra logging, and it looks like I'm just impatient. It's successfully looking up the DNS on the IPv4 DNS server and then failing to do so against the IPv6 server (and there are, in fact, no IPv6 packets in Wireshark). Because I have a bunch of names, it'll take ten minutes to wait out all the IPv6 timeouts, and I never did.
> 
> I'll try to debug my IPv6 connectivity.

Yeah. Would be good to know if it's the VPN admins' fault (DNS-over-IPv6 is configured wrong on the server side) or `vpn-slice`'s fault (traffic to DNS servers' IPv6 addresses is being routed incorrectly).

> I think it's surprising that, after successfully resolving a name against one name server, it tries again against the other. It strikes me that a configuration where AAAA records are only returned over IPv6 would be a very unusual one.

Unfortunately there are DNS servers and VPNs which are known to misbehave in ways like this, probably as a crutch for older client OSes that themselves mishandled IPv6.

It's not _exactly_ the same thing, but Pulse VPNs won't let you tunnel IPv6 packets to the VPN when connected to the _gateway_ via IPv4, and vice versa… a completely idiotic arrangement that defeats the entire purpose of a _tunneling_ protocol, but [nevertheless true](https://gitlab.com/openconnect/openconnect/-/commit/b4f50f8bd5da7e6ac926ddd5095501edbc204cd0).

We _could_ change the `keep_going` parameter to default to `False`, and ensure that IPv4 servers are tried first… but then you would fail to get IPv6 addresses for VPNs that only send `A` records via DNS-over-IPv4.

**tl;dr** It seems to me that the bottom line here is we need to figure out why IPv6 connectivity to your DNS servers isn't working, and fix or workaround that, rather than simply ignore or skip DNS-over-IPv6 silently. (We could also add an `--ignore-ipv6-dns` option to `vpn-slice` itself for cases where the IPv6 DNS server addresses are intractably broken and the VPN admins won't fix.)

**Comment by ciapecki** [comment]
> > it works but still way too long. I tested for 5 servers it took 2,5 minutes to resolve.
> > on v0.14 it works almost immediately (couple of seconds)
> 
> Hmmm… 2.5 minutes is impossibly long with a 30 second timeout. There are at most 4 combinations of (DNS server address: IPv4 vs. IPv6 ) × (DNS rectype: `A` vs `AAAA`).
> 
> @ciapecki Can you share the logging output? It should not take more than 2 minutes given 4 combinations and a 30 second timeout.

I have 5 hosts passed to vpn-slice. For every one it takes 30seconds => together ~2,5min

for every host:
```
Got results: [<DNS IN A rdata: AAA>]
[30seconds wait time]
  login.myserver.com = AAA
```

**Comment by dlenski** [comment]
> I have 5 hosts passed to vpn-slice. For every one it takes 30seconds => together ~2,5min

Ah, thanks for clarifying. :ok_hand: 

> for every host:
> ```
> Got results: [<DNS IN A rdata: AAA>]
> [30seconds wait time]
>   login.myserver.com = AAA
> ```

Yeah, this has to be because of a timeout on whichever type of DNS server it is trying in the _first_ 30 seconds.

Can you please try 505e136 (= fdf9ef2 with the logging statements uncommented)? This will show you whether or not it's the DNS-over-IPv6 or DNS-over-IPv4 queries which are timing out; that's an important piece of information to help us know whether this is really an IPv6-specific problem.

**Comment by ciapecki** [comment]
> > I have 5 hosts passed to vpn-slice. For every one it takes 30seconds => together ~2,5min
> 
> Ah, thanks for clarifying. 
> 
> > for every host:
> > ```
> > Got results: [<DNS IN A rdata: AAA>]
> > [30seconds wait time]
> >   login.myserver.com = AAA
> > ```
> 
> Yeah, this has to be because of a timeout on whichever type of DNS server it is trying in the _first_ 30 seconds.
> 
> Can you please try [505e136](https://github.com/dlenski/vpn-slice/commit/505e136b6aa7737b3d85c500b847deb608ab3721) (= [fdf9ef2](https://github.com/dlenski/vpn-slice/commit/fdf9ef2bcae634cbadace905df569f17b528f7b0) with the logging statements uncommented)? This will show you whether or not it's the DNS-over-IPv6 or DNS-over-IPv4 queries which are timing out; that's an important piece of information to help us know whether this is really an IPv6-specific problem.

```
WARNING: IPv6 address or netmask set. Support for IPv6 in vpn-slice should be considered BETA-QUALITY.
WARNING: IPv6 address or netmask set. Support for IPv6 in vpn-slice should be considered BETA-QUALITY.
Blocked incoming traffic from VPN interface with iptables.
Added routes for 3 nameservers, 0 subnets, 0 aliases.
Restored routes for 0 excluded subnets. []
Adding /etc/hosts entries for 3 nameservers...
  XXX = dns0.tun0
  YYY = dns1.tun0
  ZZZZZZ = dns2.tun0
Looking up 5 hosts using VPN DNS servers...
Issuing query for hostname 'my.mycompany.com', rectype 'A', source IPv4Address('AAA'), search_domains [<DNS name nl.mycompany.com.>], nameservers ['XXX', 'YYY']
Got results: [<DNS IN A rdata: BBB>]
Issuing query for hostname 'my.mycompany.com', rectype 'AAAA', source IPv4Address('AAA'), search_domains [<DNS name nl.mycompany.com.>], nameservers ['XXX', 'YYY']
Issuing query for hostname 'my.mycompany.com', rectype 'A', source IPv6Interface('SSSSSS/64'), search_domains [<DNS name nl.mycompany.com.>], nameservers ['ZZZZZZ']


30secs....

  my.mycompany.com = BBB
Issuing query for hostname 'blog.us.mycompany.com', rectype 'A', source IPv4Address('AAA'), search_domains [<DNS name nl.mycompany.com.>], nameservers ['XXX', 'YYY']
Got results: [<DNS IN A rdata: CCC>]
Issuing query for hostname 'blog.us.mycompany.com', rectype 'AAAA', source IPv4Address('AAA'), search_domains [<DNS name nl.mycompany.com.>], nameservers ['XXX', 'YYY']
Issuing query for hostname 'blog.us.mycompany.com', rectype 'A', source IPv6Interface('SSSSSS/64'), search_domains [<DNS name nl.mycompany.com.>], nameservers ['ZZZZZZ']

30secs....

  blog.us.mycompany.com = CCC

Issuing..... for other 3 hosts

Added hostnames and aliases for 8 addresses to /etc/hosts.
Added 5 routes for named hosts.
Connection setup done, child process 221281 exiting.
```

**Comment by dlenski** [comment]
Thanks @ciapecki. Okay, we can definitely confirm from this that it is indeed the DNS-over-IPv6 servers which are timing out for you

**Comment by fahasch** [comment]
I have an issue that I think is related. For me vpn-slice stopped working at version v0.14.1. In particular, I get the error:

`Traceback (most recent call last):
  File "/usr/local/bin/vpn-slice", line 8, in <module>
    sys.exit(main())
  File "/usr/local/Cellar/vpn-slice/0.14.1/libexec/lib/python3.8/site-packages/vpn_slice/__main__.py", line 509, in main
    do_connect(env, args)
  File "/usr/local/Cellar/vpn-slice/0.14.1/libexec/lib/python3.8/site-packages/vpn_slice/__main__.py", line 168, in do_connect
    providers.route.add_address(env.tundev, env.myaddr6)
  File "/usr/local/Cellar/vpn-slice/0.14.1/libexec/lib/python3.8/site-packages/vpn_slice/mac.py", line 109, in add_address
    self._ifconfig(device, family, ip_network(address), address)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/ipaddress.py", line 79, in ip_network
    return IPv6Network(address, strict)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/ipaddress.py", line 2106, in __init__
    raise ValueError('%s has host bits set' % self)
ValueError: ZZZZZZZZ/64 has host bits set`

If I look at the dump and compare version 0.14.0 to version 0.14.1, I notice that
INTERNAL_IP6_ADDRESS has changed. In version 0.14.0 it has been simply "ZZZZZZZZ" (without any subnet). In version 0.14.1, it is "ZZZZZZZZ/64". This problem seems to be related to the one reported by @ciapecki . The variable INTERNAL_IP6_NETMASK has a subnet in both versions.

Note however that I do not  have timing out issues but the error that "... has host bits set".

**Comment by dlenski** [comment]
Merged #55 from which should resolve the issue on both macOS and Linux.

Brief summary is that `dnspython` does not tolerate the "interface" form of an IPv4/6 source address which combines the host address and netmask (e.g. `1.2.3.4/24` or `2601:f00f::f00b/64`).

This affected IPv6 due to the changes to store the IPv6 address in this form between `v0.14` and `v0.14.1` (needed for correct behavior with IPv6 _routing_). I didn't test the DNS-over-IPv6 interaction properly _in this form_ because I haven't had access to a VPN with IPv6 DNS servers in the last few weeks (bad me :man_facepalming: :-1:).

Three cheers to @gmacon for figuring out a solution and writing a clean fix. :beers:
